File: 03-backtracking-generation/generation-1.py
def init(b, n):
    for i in range(n):
        b.append(0)

def out(b, n):
    print("".join(map(str, b[:n])))

def islast(b, n):
    return all(x == 1 for x in b[:n])

def gen(b, n):
    i = n - 1
    while i >= 0 and b[i] == 1:
        i -= 1
    if i >= 0:
        b[i] = 1
        for j in range(i + 1, n):
            b[j] = 0

def method(b, n):
    init(b, n)
    out(b, n)
    while not islast(b, n):
        gen(b, n)
        out(b, n)

def main():
    n = 4
    b = []
    method(b, n)

if __name__ == "__main__":
    main()

File: 03-backtracking-generation/generation-1.cpp
#include <iostream>
using namespace std;

void init(int b[], int n) {
    for (int i = 0; i < n; i++) b[i] = 0;
}

void out(int b[], int n) {
    for (int i = 0; i < n; i++) cout << b[i] << " ";
    cout << endl;
}

int islast(const int b[], int n) {
    for (int i = 0; i < n; i++) if (b[i] == 0) return 0;
    return 1;
}

void gen(int b[], int n) {
    int i = n - 1;
    while (i >= 0 && b[i] == 1) i--;
    if (i >= 0) {
        b[i] = 1;
        for (int j = i + 1; j < n; j++) b[j] = 0;
    }
}

void method(int b[], int n) {
    init(b, n);
    out(b, n);
    int stop = islast(b, n);
    while (stop == 0) {
        gen(b, n);
        out(b, n);
        stop = islast(b, n);
    }
}

int main() {
    int n = 4;
    int b[n];
    method(b, n);
    return 0;
}